File: week5/diamond.cpp
#include<iostream>
using namespace std;
class Demo1{
    public: 
        int a=10;
        void demo1Msg(){
            cout<<"Demo 1 class\n";
        }
};
//simple inheritance
class Demo2 :virtual public Demo1{
    public:
    void demo2Msg(){
            cout<<"Demo 2 class\n";
        }
};
//simple inheritance
class Demo3: virtual public Demo1{
    public:
    void demo3Msg(){
            cout<<"Demo 3 class\n";
        }
};
//Multiple inheritance
class Demo4: virtual public Demo2, virtual public Demo3{
    public:
    void demo4Msg(){
            cout<<"Demo 4 class\n";
        }
};
int main(){
    Demo4 obj;
    cout<<"Diamond inheritance\n";
    cout<<obj.a<<endl;//ambiguous
    obj.demo1Msg();//ambiguous
    obj.demo2Msg();
    obj.demo3Msg();
    obj.demo4Msg();
}

File: week5/Diamond.java
// class Demo1{
//     int a=10;
//     void display1(){
//         System.out.println("Demo class 1");
//     }
// }
// class Demo2 extends Demo1{
//     int b=10;
//     void display2(){
//         System.out.println("Demo class 2");
//     }
// }
// class Demo3 extends Demo1{
//     int c=10;
//     void display3(){
//         System.out.println("Demo class 3");
//     }
// }
// class Demo4 extends Demo2 , Demo3{
//     int b=10;
//     void display4(){
//         System.out.println("Demo class 4");
//     }
// }
// class Diamond{
//     public static void main(String[] args) {
//         Demo4 obj=new Demo4();
//         obj.display1();
//     }
// }

interface DemoInterface1  
{  
public default void display()   
{  
System.out.println("the display() method of DemoInterface1 invoked");  
}  
}  
interface DemoInterface2  
{  
public default void display()   
{  
System.out.println("the display() method of DemoInterface2 invoked");  
}  
}  
public class Diamond implements DemoInterface1, DemoInterface2  
{  
public void display()   
{  
DemoInterface1.super.display();  
DemoInterface2.super.display();  
}  
public static void main(String args[])   
{  
Diamond obj = new Diamond();  
obj.display();  
}  
}  